src/toil/leader.py

```diff
@@ -583,29 +586,54 @@ def _processReadyJob(self, job_id: str, result_status: int):
             # Use the service manager to start the services
             self.serviceManager.put_client(job_id)
         elif len(readyJob.stack) > 0:
-            # There are exist successors to run
+            # There are successors to run
             self._runJobSuccessors(job_id)
         elif readyJob.jobStoreID in self.toilState.servicesIssued:
             logger.debug("Telling job: %s to terminate its services due to the "
                          "successful completion of its successor jobs",
                          readyJob)
             self.serviceManager.kill_services(self.toilState.servicesIssued[readyJob.jobStoreID], error=False)
         else:
-            #There are no remaining tasks to schedule within the job, but
-            #we schedule it anyway to allow it to be deleted. Remove the job
+            # There are no remaining tasks to schedule within the job.
+            #
+            # We want it to go away. Usually the worker is responsible for
+            # deleting completed jobs. But when the job has a
+            # child/followOn/service dependents that have to finish first, it
+            # gets kicked back here when they are done.
+            #
+            # We don't want to just issueJob() again because then we will
+            # forget that this job is actually done and not a retry. A
+            # successful-subtree checkpoint job can't be distinguished from a
+            # being-retried one.
+            #
+            # So we do the cleanup here, to avoid reserving any real resources.
+            # See: https://github.com/DataBiosphere/toil/issues/4158 and
+            # https://github.com/DataBiosphere/toil/issues/3188
+            #
+            # TODO: Does this want to happen in another thread? What if we
+            # don't manage to do all the jobstore file deletes and need to
+            # retry them? What if there are 1000 of them?
```

Comment:
I think this way is fine. In practice, do you think this would be a problem when we have large numbers of job-related files that need to be cleaned up?

Reply:
I think it could start to be a problem when the files are on high-latency, remote storage like S3. If your ping to S3 is like 30 ms, and you have 3000 files, deleting them all in a row might take 90000 ms or 90 seconds, even if you are pooling HTTP connections so you don't have to reconnect every time. And during that time the leader won't be doing things like scheduling new jobs to replace other jobs that finish, or updating the progress bar.